### DISCUSSION TITLE
Float numbers as input names

### DISCUSSION THREAD

**Discussion by ghost** [react-hook-form/react-hook-form#3454] [Q&A] [answered]
What's up. I'm trying to create a `form` which has many `radio`s and on each `radio` I use the number of the question as its `name`,  this is a requirement.

When I try this:

```js
 <section>
        <label>Question?</label>

        <Controller
          as={
            <RadioGroup>
              <FormControlLabel value={true} control={<Radio />} label="Yes" />

              <FormControlLabel value={false} control={<Radio />} label="Not" />
            </RadioGroup>
          }
          name="1"
          control={methods.control}
          defaultValue={false}
        />

        <Typography className={classes.error}>
          <ErrorMessage errors={methods.errors} name="1" />
        </Typography>
</section>

 <section>
        <label>Question?</label>

        <Controller
          as={
            <RadioGroup>
              <FormControlLabel value={true} control={<Radio />} label="Yes" />

              <FormControlLabel value={false} control={<Radio />} label="Not" />
            </RadioGroup>
          }
          name="1.1"
          control={methods.control}
          defaultValue={false}
        />

        <Typography className={classes.error}>
          <ErrorMessage errors={methods.errors} name="1.1" />
        </Typography>
</section>
```

My output looks like `{1: [empty, false ] }`, while I need something like `{1: false, '1.1': false }`. How can I fix the problem of empty value, that I don't have any idea of where it is coming (my form is extremely simple, I think React Hook Form is generating this empty value), and the `name` -> key problem?

React Hook Form version 6.11.0.

**Comment by bluebill1049** [comment]
hey there :)  you want to try this: https://react-hook-form.com/advanced-usage#TransformandParse

**Reply by ghost** [reply]
Is there a way to do this "globally", the section of the platform that I'm working has 10 different forms and radio, checkbox, text, etc; as input types. I just need to get all the keys / values as separated sets. If I do this manually it'll be a pain in the ass.

**Reply by bluebill1049** [reply]
short answer: no. you will have to write your own parser.

good news, we consider to include valueAsNumber for `register` as first step.

**Reply by ghost** [reply]
Same problem over and over [again](https://spectrum.chat/react-hook-form/help/yes-no-radios-as-boolean-true-false~c4b15b76-2c46-4d4b-8038-ac132cb87206).

Good to know that the contributors are considering, but for now I'm suffering. I'll just map everything on `onSubmit`, its has been a huge (real) headache (like it was when we used Formik with components and not hooks) to parse all these input types and get the right type, e.g `Boolean`, on `onSubmit`.

**Reply by ghost** [reply]
I have never contributed to this library and I know that you work really hard on it (thanks for doing it), but holy shit, when it gets to something different from the basics it gets insanely hard (or time-consuming) to make things work the way you expect them to work, like the case that I'm working right now.

Observation: The previous developers from another project that I'm trying to keep alive hate this library, cause the instability on the uncontrolled form f****d them really hard, to the point that they made the shit decision of add another library that does the same thing just to not deal with React Hook Forms. I'm starting to understand why they made this shit decision, I had an idea before when I needed to read the changelogs, documentation, threads on STF, Github, Spectrum, and other sites, Proxy, Observable, Generator and other APIs to fix extremely hard bugs on considerable large forms involving uncontrolled forms.

This comment looks like a product review and not a standard issue feedback, but yeah, just ignore it. : )

**Reply by bluebill1049** [reply]
> cause the instability on the uncontrolled form f****d them really hard.

what's an example of usage? 

look obviously I can hear your frustration and pain, but what is this have to do with the uncontrolled form? your problem is mostly at the input/field level, which is input name then moving into correct data type is not cast, I find it hard to understand the difficulty in the first place and the first time have heard this type of "extreme" feedback. 

What's the other lib that he included to not work with RHF? I would like to know, that's not the way of solving problems as well. To be clear, I am not here to win an argument, in fact, still keen to help, understand your frustration, and potential providers' solutions.

**Reply by bluebill1049** [reply]
> Is there a way to do this "globally", the section of the platform that I'm working has 10 different forms and radio, checkbox, text, etc; as input types. I just need to get all the keys / values as separated sets. If I do this manually it'll be a pain in the ass.

Just like the above, Do you have a CSB? otherwise, it's difficult to even imagine the use case.

**Reply by bluebill1049** [reply]
> Same problem over and over [again](https://spectrum.chat/react-hook-form/help/yes-no-radios-as-boolean-true-false~c4b15b76-2c46-4d4b-8038-ac132cb87206).
> 
> Good to know that the contributors are considering, but for now I'm suffering. I'll just map everything on `onSubmit`, its has been a huge (real) headache (like it was when we used Formik with components and not hooks) to parse all these input types and get the right type, e.g `Boolean`, on `onSubmit`.

This question is answered at Spectrum: (repeat it here)

native radio input doesn't return boolean, this web standard. However, you can build a controlled input and wrapped with Controller. here is an example below:

https://codesandbox.io/s/unruffled-wing-r15gt?file=/src/App.js

**Reply by ghost** [reply]
> > cause the instability on the uncontrolled form f****d them really hard.
> 
> what's an example of usage?
> 
> look obviously I can hear your frustration and pain, but what is this have to do with the uncontrolled form? your problem is mostly at the input/field level, which is input name then moving into correct data type is not cast, I find it hard to understand the difficulty in the first place and the first time have heard this type of "extreme" feedback.
> 
> What's the other lib that he included to not work with RHF? I would like to know, that's not the way of solving problems as well. To be clear, I am not here to win an argument, in fact, still keen to help, understand your frustration, and potential providers' solutions.

The uncontrolled form was in another project, not the one that I'm working on right now, which is the one that originated this entire question.

The problem at that case was that:

1. The `watch` returned an empty state between renders and even if I verified that it was empty and set a default state it didn't work, i.e for `watch` the state was still empty

2. Cause the inputs were registered using `setField`, we needed to re-register them setting default values based on specific changes on the interface, conditional inputs, and cause `watch` returned an empty state - the previously registered data was not shown on the inputs

3. When removing an item from the uncontrolled form state it caused a strange behavior that cleared all the data

4. Cause all the data was cleared on the uncontrolled form state, but it still on the interface when the user tried to remove an item the page broke - internally on RHF it was not possible to iterate on an `undefined` value

There was more, but this gives a good idea of what I'm talking about. Update the library fixed a few bugs, but of course, it had taken a huge time to make all the code compatible everywhere.

**Reply by ghost** [reply]
> > Same problem over and over [again](https://spectrum.chat/react-hook-form/help/yes-no-radios-as-boolean-true-false~c4b15b76-2c46-4d4b-8038-ac132cb87206).
> > Good to know that the contributors are considering, but for now I'm suffering. I'll just map everything on `onSubmit`, its has been a huge (real) headache (like it was when we used Formik with components and not hooks) to parse all these input types and get the right type, e.g `Boolean`, on `onSubmit`.
> 
> This question is answered at Spectrum: (repeat it here)
> 
> native radio input doesn't return boolean, this web standard. However, you can build a controlled input and wrapped with Controller. here is an example below:
> 
> https://codesandbox.io/s/unruffled-wing-r15gt?file=/src/App.js

Thanks, I have read this yesterday. I made it work with Material UI (radio, checkbox, select), `Boolean` values and setting the data from the API. Just want to finish fast ( :

Observation: Theoretically none of the links that I found, even the one that you sent at this thread, fix the problem with the keys, even the parse from the documentation. Let's just change the keys to something like `1-1`, `1-2` and parse it later.

**Reply by ghost** [reply]
> > Is there a way to do this "globally", the section of the platform that I'm working has 10 different forms and radio, checkbox, text, etc; as input types. I just need to get all the keys / values as separated sets. If I do this manually it'll be a pain in the ass.
> 
> Just like the above, Do you have a CSB? otherwise, it's difficult to even imagine the use case.

Don't have access to the project anymore.

**Reply by bluebill1049** [reply]
> 1. The `watch` returned an empty state between renders and even if I verified that it was empty and set a default state it didn't work, i.e for `watch` the state was still empty

Don't think this is the case until you can bring a reproducible CSB, watch has no magic, it's the native form behavior, the value is gong to be empty before hit tot he renders, that's why even with controlled input, you will need to set up the initial state. The whole thing is just native JS.

**Reply by bluebill1049** [reply]
> 2\. Cause the inputs were registered using `setField`, we needed to re-register them setting default values based on specific changes on the interface, conditional inputs, and cause `watch` returned an empty state - the previously registered data was not shown on the inputs

Not 100% sure the usage here, you can always register an item with a custom register which is at useEffect, treat it as virtual input.

**Reply by bluebill1049** [reply]
> 3\. When removing an item from the uncontrolled form state it caused a strange behavior that cleared all the data

This is native input behavior if you remove an input from the DOM, the input value will be lost. If you want to persist your data, use `shouldUnregister: false` or simplify just hide the visibility. We tend to love mount and unmount components at React, and forgot how simple and easy things were to show and hide an iteam.

**Reply by bluebill1049** [reply]
> 4\. Cause all the data was cleared on the uncontrolled form state, but it still on the interface when the user tried to remove an item the page broke - internally on RHF it was not possible to iterate on an `undefined` value

Need more context and example.

**Reply by bluebill1049** [reply]
> There was more, but this gives a good idea of what I'm talking about. Update the library fixed a few bugs, but of course, it had taken a huge time to make all the code compatible everywhere.

That's the thing, doesn't matter which lib that you pick, you have to work with it. They all have their own design decision (cons and props), if you work against it, no matter how much time you spent it will not be efficient, I don't see those questions above are a show-stopper, consider all those use cases that we have them all at work (we work on forms every day at work with this lib). I would suggest you think RHF more of a native form lib, I think most of the roadblock and your headache will ease up.

**Reply by ghost** [reply]
Yeah, anyway. Made it work, thanks for the support. ^ - ^

**Reply by bluebill1049** [reply]
I feel bad to hear your frustration... please don't hesitate to ask questions again. I am here to support everyone.

**Comment by bluebill1049** [comment, marked as answer]
Ok, looks like i didn't read your first question carefully. the answer is in the doc:

> Important: name is required and unique (can not start with a number or use number as key name). Input name also supports dot and bracket syntax, which allows you to easily create nested form fields.

<img width="946" alt="Screen Shot 2020-11-22 at 5 56 08 pm" src="https://user-images.githubusercontent.com/10513364/99897238-01db8600-2cec-11eb-9ed2-d280d427bac1.png">

**Reply by ghost** [reply]
Yeah, I also read it. The logic of the API is based on numbers to identify resources from sections / sub-sections, that's why I needed to make it work with numbers. The `register` documentation doesn't apply to the radio, checkbox and select inputs, which are the core of my forms, cause to integrate it with Material UI I need to use [Controller](https://react-hook-form.com/api#Controller).
